### PR TITLE
[Modelzoo]Rebuild DBMTL to update API and Enable DeepRec Features

### PR DESCRIPTION
**Pull request by jmaksymc**
**WIP**

- Enable following DeepRec feature:
  - Enabled By Default:
      8) ~~Auto Micro Batch same with https://github.com/alibaba/DeepRec/issues/127~~
      9) ~~FusedEmbedding API, embedding fusion~~
      10) ~~Smart Stage same with https://github.com/alibaba/DeepRec/issues/122~~
      11) ~~Auto Graph Fusion https://github.com/alibaba/DeepRec/issues/144~~
      12) ~~CPU Memory Optimization:START_STATISTIC_STEP, STOP_STATISTIC_STEP, jemalloc~~
      14) ~~AdamAsync Optimizer~~
      15) ~~BF16~~
  - Disabled By Default:
         1) Embedding Variable **Not support EV yet**
         7) GRPC++ and StarServer **Implemented in the model but they are not supported yet**
         13) ~~Incremental Checkpoint~~
         14) ~~AdagradDecay~~
         2) EmbeddingVariable advanced features：Embedding Elimination **Not support EV yet**
         3) EmbeddingVariable advanced feature：Embedding Filter **Not support EV yet**
         4) Dynamic-dimension Embedding Variable - **There is no feature_column API**
         5) Adaptive Embedding - **ERROR:**  data.shape must start with partitions.shape, got data.shape = [18886], partitions.shape = [512]
          17) ~~WorkQueue~~
  - Other Features:
      6) Multi-Hash Variable - **There is no feature_column API**

### CODE REVIEW

File: modelzoo/DBMTL/DockerFile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Dockerfile for DLRM training
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/DockerFile:1]: Please remove DockerFile

File: modelzoo/DBMTL/train.py [hunk 2]
```diff
@@ -0,0 +1,921 @@
...
+        raise ValueError('Not a valid boolean string')
+    return low_string == 'true'
+
+def get_arg_parser():
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    # parser.add_argument('--job',
+    #                     help='train or test model',
+    #                     type=str,
+    #                     choices=["train", "test"],
+    #                     default='train')
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:710]: Clean up comments

Comment (jmaksymc) [reply modelzoo/DBMTL/train.py]: Cleaned up unnecessary comments, there are still some that I need for WIP.

**Comment by jmaksymc** [comment]
Fixed smartstaged feature after rebasing to main and building DeepRec again.

File: modelzoo/DBMTL/train.py [hunk 3]
```diff
@@ -0,0 +1,900 @@
...
+            self._create_metrics()
+    
+    def _add_layer_summary(self, value, tag):
+        tf.summary.scalar('%s/fraction_of_zero_values' % tag,
+                        tf.nn.zero_fraction(value))
+        tf.summary.histogram('%s/activation' % tag, value)
+    
+    def _make_scope(self, name, bf16, part):
+        if(bf16):
+            return tf.variable_scope(name, partitioner=part, reuse=tf.AUTO_REUSE).keep_weights()
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:149]: keep_weights(dtype=tf.float32), this makes it clearer what the function does.

File: modelzoo/DBMTL/train.py [hunk 4]
```diff
@@ -0,0 +1,900 @@
...
+    else:
+        files = filename
+    # Extract lines from input files using the Dataset API.
+    dataset = tf.data.TextLineDataset(files)
+    dataset = dataset.shuffle(buffer_size=20000,
+                              seed=args.seed)  # set seed for reproducing
+    dataset = dataset.repeat(num_epochs)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(parse_csv, num_parallel_calls=28)
+    dataset = dataset.prefetch(1)
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:349]: prefetch(2), keep aligned with other models in modelzoo.

File: modelzoo/DBMTL/train.py [hunk 5]
```diff
@@ -0,0 +1,900 @@
...
+    print("Number of training dataset is {}".format(no_of_training_examples))
+    print("Number of test dataset is {}".format(no_of_test_examples))
+
+    # set batch size, epoch & steps
+    batch_size = math.ceil(
+        args.batch_size / args.micro_batch
+    ) if args.micro_batch and not args.tf else args.batch_size
+
+    if args.steps == 0:
+        no_of_epochs = 3
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:597]: Please be sure that by python train.py, the model will be trained for at least 3000 steps. Because the test script in DeepRec will run python train.py and collect the gsteps between 2000~3000. The script is under DeepRec/tests/model_benchmark.

File: modelzoo/DBMTL/train.py (hunk 4 above)
Comment (jmaksymc) [reply modelzoo/DBMTL/train.py]: Changed

File: modelzoo/DBMTL/train.py (hunk 5 above)
Comment (jmaksymc) [reply modelzoo/DBMTL/train.py:597]: Changed

**Comment by Duyi-Wang** [comment]
Please update README、distribute training files and add DBMTL in modelzoo/README. Adaptive Embedding issue has been submitted to Ali https://github.com/alibaba/DeepRec/issues/236

File: modelzoo/DBMTL/train.py (hunk 5 above)
Comment (Duyi-Wang) [reply modelzoo/DBMTL/train.py:597]: Please change deafult steps by `no_of_epochs`

**Comment by shanzhou2186** [comment]
Please help to change the steps number and update the performance data of Alicloud CPX in JIRA and ignore the accuracy for such a small dataset. Then we can push the PR to Alibaba.

**Comment by Duyi-Wang** [comment]
- Please add a directory structure at the start of REAMDE.
- Please update distribute training files, keep align with other models.

**Comment by jmaksymc** [comment]
Updated, performance is on JIRA.

File: modelzoo/DBMTL/train.py [hunk 6]
```diff
@@ -0,0 +1,900 @@
...
+                    [model.acc_op, model.auc_op, merged])
+                writer.add_summary(events, _in)
+                print("Evaluation complete:[{}/{}]".format(_in, steps))
+                print("ACC = {}\nAUC = {}".format(eval_acc, eval_auc))
+
+def main(tf_config=None, server=None):
+    # check dataset and count data set size
+    print("Checking dataset...")
+    train_file = os.path.join(args.data_location, 'train')
+    test_file = os.path.join(args.data_location, 'eval')
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:581]: train_file = os.path.join(args.data_location, 'taobao_train_data')
    test_file = os.path.join(args.data_location, 'taobao_test_data')

File: modelzoo/DBMTL/train.py [hunk 7]
```diff
@@ -0,0 +1,900 @@
...
+            #         '''
+            #         column = tf.feature_column.categorical_column_with_adaptive_embedding(
+            #             column_name,
+            #             hash_bucket_size=HASH_BUCKET_SIZES[column_name],
+            #             dtype=tf.string,
+            #             ev_option=ev_opt)
+            #     elif args.dynamic_ev:
+            #         '''Dynamic-dimension Embedding Variable'''
+            #         print("Dynamin-dimension Embedding Variable isn't really enabled in model.")
+            #         sys.exit()
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:475]: clean

File: modelzoo/DBMTL/train.py [hunk 8]
```diff
@@ -0,0 +1,900 @@
...
+'''
+
+class DBMTL():
+    def __init__(self,
+                 input,
+                 feature_column,
+                 bottom_dnn,
+                 towers,
+                 relation_dnn,
+                #  l2_scale,
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:107]: clean

File: modelzoo/DBMTL/train.py [hunk 9]
```diff
@@ -0,0 +1,900 @@
...
+        if not input:
+            raise ValueError("Dataset is not defined.")
+        self._feature = input[0]
+        self._label = input[1]
+        
+        self._feature_column = feature_column
+        self._bottom_dnn = bottom_dnn
+        self._towers = towers
+        self._relation_dnn = relation_dnn
+        # self.__l2_regularization = self._l2_regularizer(l2_scale) if l2_scale else None
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:124]: clean

File: modelzoo/DBMTL/train.py [hunk 10]
```diff
@@ -0,0 +1,900 @@
...
+                shared_features = input_emb
+                if self.bf16:
+                    shared_features = tf.cast(shared_features, dtype=tf.bfloat16)
+
+                for layer_id, num_hidden_units in enumerate(self._bottom_dnn):
+                    with self._make_scope(f'bottom_dnn_layer_{layer_id}', self.bf16, self._dense_layer_partitioner) as shared_layer_scope:
+                        shared_features = tf.layers.dense(shared_features,
+                                                        units=num_hidden_units,
+                                                        activation=None,
+                                                        # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:194]: clean

File: modelzoo/DBMTL/train.py [hunk 11]
```diff
@@ -0,0 +1,900 @@
...
+                specific_features = tower_input
+
+                if self.bf16:
+                    specific_features = tf.cast(specific_features, dtype=tf.bfloat16)
+                    
+                for layer_id, num_hidden_units in enumerate(hidden_units):
+                    with self._make_scope(f'{tower_name}_layer_{layer_id}', self.bf16, self._dense_layer_partitioner) as tower_layer_scope:
+                        specific_features = tf.layers.dense(specific_features,
+                                                        units=num_hidden_units,
+                                                        # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:215]: clean

File: modelzoo/DBMTL/train.py [hunk 12]
```diff
@@ -0,0 +1,900 @@
...
+                                                        name=f'{relation_dnn_scope.name}/dense')
+                        relation_input = DNN_ACTIVATION(relation_input,
+                                                        name=f'{relation_dnn_scope.name}/act')
+                        self._add_layer_summary(relation_input, relation_dnn_scope.name)
+                # Dense 1
+                        final_tower_predict = tf.layers.dense(relation_input,
+                                                            units=1,
+                                                            activation=None,
+                                                            # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:243]: clean

File: modelzoo/DBMTL/train.py [hunk 13]
```diff
@@ -0,0 +1,900 @@
...
+                    for relation_name in relations[tower_name]:
+                        relation_input.append(relations_outputs[relation_name])
+
+                # Relation DNN
+                relation_input = tf.concat(relation_input, axis=1)
+                for layer_id, num_hidden_units in enumerate(self._relation_dnn):
+                    with self._make_scope(f'{tower_name}_relation_dnn_{layer_id}', self.bf16, self._dense_layer_partitioner) as relation_dnn_scope:                      
+                        relation_input = tf.layers.dense(relation_input,
+                                                        units=num_hidden_units,
+                                                        # kernel_regularizer=self.__l2_regularization,
```
Comment (Duyi-Wang) [inline modelzoo/DBMTL/train.py:234]: clean

**Comment by jmaksymc** [comment]
Cleaned comments and changed train/test files names.
